File: 2022/haskell/04/Main.hs
module Main (main) where

import Helpers
import Data.List.Split (splitOn)

stringToAssignmentPair :: String -> AssignmentPair
stringToAssignmentPair s = ((a1 !! 0, a1 !! 1), (a2 !! 0, a2 !! 1))
    where
        stringByElf = splitOn [','] s
        stringByElfToAssignment = (\i -> map read $ splitOn ['-'] (stringByElf !! i))
        a1 = stringByElfToAssignment 0 -- [2, 4]
        a2 = stringByElfToAssignment 1 -- [6, 8]

main :: IO()
main = do
    inputFile <- readFile "../../inputs/04.txt"
    let overlaps = map getOverlap $ map stringToAssignmentPair $ lines inputFile
    let countOverlaps = (\f -> length $ filter f overlaps)
    let result1 = countOverlaps (\overlap -> overlap == Full)
    let result2 = countOverlaps (\overlap -> overlap == Full || overlap == Partial)
    print result1
    print result2

File: 2022/python/04/main.py
result_1 = 0
result_2 = 0

with open('../../inputs/04.txt', 'r') as input_file:
    for line in input_file:
        l, r = line.strip().split(',')
        l1, l2 = map(int, l.split('-'))
        r1, r2 = map(int, r.split('-'))
        if (l1 <= r1 and l2 >= r2) or (l1 >= r1 and l2 <= r2): # full overlap
            result_1 += 1
            result_2 += 1
        elif (l1 <= r1 <= l2) or (l1 <= r2 <= l2): # partial overlap
            result_2 += 1

print(result_1)
print(result_2)
